### DISCUSSION TITLE
Session Flash not clearing when in a layout

### DISCUSSION THREAD

**Discussion by kunanyi** [remix-run/remix#9679] [Q&A]
I'm having issues getting Flash messaging to work in a layout.

In [this repo](https://github.com/kunanyi/sess) I've done a `session.get` in the root, as a universal spot to show any flash error messages.

**Problem is**, the session flash does not clear.  It persists.

If I move the `session.get` from the root layout to a lower layout (like "movie"), it's the same issue.   It only seems to work if I put it inside a route endpoint like `routes/movie.index`?

Any suggestions on how to get the Flash display to work (and clear) from a universal location?

Thanks 😊

--------
--------

The [repo](https://github.com/kunanyi/sess) is just a plain remix install.  

It uses [app/sessions.ts](https://remix.run/docs/en/main/utils/sessions) from the docs.

With these two files the main difference:

### /app/root.tsx

```
import { Links, Meta, Outlet, Scripts, ScrollRestoration, json, useRouteLoaderData } from "@remix-run/react";
import "./tailwind.css";
import { commitSession, getSession } from "./sessions";
import { LoaderFunctionArgs } from "@remix-run/cloudflare";

export async function loader({ request }: LoaderFunctionArgs) {
	const session = await getSession(request.headers.get("Cookie"));
	const error = session.get("error");
	console.log("loading");
	return json(
		{ error },
		{
			headers: {
				"Set-Cookie": await commitSession(session),
			},
		}
	);
}

export function Layout({ children }: { children: React.ReactNode }) {
	/* tsignore-next-line */
	const { error } = useRouteLoaderData("root") as any;
	return (
		<html lang="en">
			<head>
				<meta charSet="utf-8" />
				<meta
					name="viewport"
					content="width=device-width, initial-scale=1"
				/>
				<Meta />
				<Links />
			</head>
			<body className="m-20">
				<p>Flash error: {error || "None"}</p>
				<br />
				{children}
				<ScrollRestoration />
				<Scripts />

				<p className="mt-12 text-xs text-gray-500">Epoch time just to see root updating: {Date.now()}</p>
			</body>
		</html>
	);
}

export default function App() {
	return <Outlet />;
}
```

### app/routes/movie.$movieId.edit.tsx

```
import { Form } from "@remix-run/react";
import { getSession, commitSession } from "../sessions";
import { ActionFunctionArgs, json, LoaderFunctionArgs, redirect } from "@remix-run/cloudflare";

export async function action({ request }: ActionFunctionArgs) {
	const session = await getSession(request.headers.get("Cookie"));
	session.flash("error", "Flash error from edit action");
	return redirect("/movie", {
		headers: {
			"Set-Cookie": await commitSession(session),
		},
	});
}

export async function loader({ request }: LoaderFunctionArgs) {
	return json({});
}

export default function Index() {
	return (
		<>
			<h1 className="text-3xl">Edit Movie</h1>
			<Form method="POST">
				<button
					className="underline"
					type="submit">
					Save
				</button>
			</Form>
			<a className="underline" href="javascript:history.back()">Back</a>
		</>
	);
}
```

**Comment by kunanyi** [comment]
OK so it seems that parent loaders aren't called unless there is a mutation for the current parent route.

So perhaps this is the issue?  After a movie is edited, the flash is retrieved by the loader.... but then that state (with the flash value present) is never cleared unless you change route entirely or cause a mutation?

If so, any suggestions for displaying flash in a global location?  Or is it too tricky with remix caches/nesting?

**Reply by do4Mother** [reply]
just add useEffect in root.tsx with timeout to clear error message
